ossm: add script for uploading built ztunnel to gs

zmiklank opened the pull request:
This should be similar to https://github.com/openshift-service-mesh/proxy/blob/release-1.27/ossm/ci/post-submit.sh and should have the same purpose.

I have not done any steps to create `gs://maistra-prow-testing/ztunnel` yet.

zmiklank committed:
ossm: add script for uploading built ztunnel to gs

Signed-off-by: Zuzana Miklankova <zmiklank@redhat.com>

jwendell commented on ossm/ci/post-submit.sh:
```diff
@@ -0,0 +1,23 @@
+fi
+
+tar czf ./out/rust/release/ztunnel.tar.gz ./out/rust/release/ztunnel
+gsutil cp bazel-bin/envoy_tar.tar.gz "${ARTIFACTS_GCS_PATH}/ztunnel-alpha-${SHA}${ARCH_SUFFIX}.tar.gz"
```
1) we should copy ztunnel, not envoy binary
2) are you sure we need to rename it to ztunnel-alpha? I'd check at https://github.com/istio/istio/blob/master/bin/build_ztunnel.sh#L127-L133

Take a look at those Istio scripts to see if it's really necessary to upload a binary artifact, or if Istio is capable of building ztunnel from source. Proxy is done that way because building proxy is really an expensive task.

zmiklank replied:
> we should copy ztunnel, not envoy binary

oh, good catch! 

> Take a look at those Istio scripts to see if it's really necessary to upload a binary artifact, or if Istio is capable of building ztunnel from source. Proxy is done that way because building proxy is really an expensive task.

Good point, will do that.
Thanks.

zmiklank replied:
Hi @jwendell. Sorry it took me that long. From what I understood from istio repo, ztunnel is, similarly to proxy, pre-built and then pushed to gcloud. So I think the approach I proposed is correct.

jwendell commented:
> I have not done any steps to create gs://maistra-prow-testing/ztunnel yet.

No need to perform any prior creation. If a directory doesn't exist, it's created on demand.

zmiklank committed:
add fixes for Jonh's findings

Signed-off-by: Zuzana Miklankova <zmiklank@redhat.com>

jwendell commented on ossm/ci/post-submit.sh:
```diff
@@ -0,0 +1,23 @@
+fi
+
+tar czf ./out/rust/release/ztunnel.tar.gz ./out/rust/release/ztunnel
+gsutil cp ./out/rust/release/ztunnel.tar.gz "${ARTIFACTS_GCS_PATH}/ztunnel-${SHA}${ARCH_SUFFIX}.tar.gz"
```
I think we don't need to tar.gz it, AFAICT upstream uses the raw binary:

https://github.com/istio/ztunnel/blob/master/scripts/release.sh#L51
and
https://github.com/istio/istio/blob/master/bin/build_ztunnel.sh#L129

plus, I think we need to add the arch to the filename: https://github.com/istio/ztunnel/blob/master/scripts/release.sh#L44

zmiklank replied:
Should be fixed now, thanks.

zmiklank committed:
ci: push raw binary, not targz, to the cloud

Signed-off-by: Zuzana Miklankova <zmiklank@redhat.com>

jwendell commented on ossm/ci/post-submit.sh:
```diff
@@ -0,0 +1,22 @@
+SHA="$(git rev-parse --verify HEAD)"
+
+if [[ "$(uname -m)" == "aarch64" ]]; then
+  ARCH_SUFFIX="-arm64"
```
remove the `-`

jwendell commented on ossm/ci/post-submit.sh:
```diff
@@ -0,0 +1,22 @@
+if [[ "$(uname -m)" == "aarch64" ]]; then
+  ARCH_SUFFIX="-arm64"
+else
+  ARCH_SUFFIX=""
```
this should be set to `amd64`: https://github.com/istio/ztunnel/blob/master/scripts/release.sh#L23

zmiklank committed:
ci: enable also x86_64 arch, fix typo

Signed-off-by: Zuzana Miklankova <zmiklank@redhat.com>

jwendell commented on ossm/ci/post-submit.sh:
```diff
@@ -0,0 +1,22 @@
+  *) echo "unsupported architecture"; exit 1;;
+esac
+
+gsutil cp ./out/rust/release/ztunnel "${ARTIFACTS_GCS_PATH}/ztunnel-${SHA}-${ARCH}"
```
Sorry @zmiklank - I just noticed - We're not performing the build anywhere in this script, thus the file `./out/rust/release/ztunnel` doesn't exist at this point.

zmiklank replied:
Build is executed in here: https://github.com/openshift/release/pull/71880/files#diff-df86c6a84809cf585a8a44b23ea657fc664ad44b456966942dcf1c95e7489bfcR54.

jwendell replied:
The way it works is: After a PR is merged in ztunnel repository, a fresh new postsubmit job is triggered (`copy-artifacts-gcs` job in that file, which basically invokes this script (line 111 on that file). Thus, this script should perform a build and the upload the binary artifact.

See the proxy cousin: https://github.com/openshift-service-mesh/proxy/blob/release-1.28/ossm/ci/post-submit.sh#L17

zmiklank committed:
WIP: adding builds where they should be

Signed-off-by: Zuzana Miklankova <zmiklank@redhat.com>

zmiklank commented:
Hi @jwendell. I added the build to the post-submit script and also added pre-submit script, which will be running ztunnel build on-PR. Could you please re-review?
This PR is prerequisite for https://github.com/openshift/release/pull/71880.

jwendell commented on ossm/ci/pre-submit.sh:
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash
+
+DIR=$(cd "$(dirname "$0")" ; pwd -P)
```
nit: variable not used.

zmiklank replied:
I am thinking, if we even need to cd to the script's directory. It seems to me that more useful could be to stay in ztunnel directory. So I will just delete the line entirely.

jwendell replied:
Note that the line above does not "enter the directory" for the purpose of running the code below it. It does that but in a subshell, only to set the `DIR` variable, which is not used anywhere.

I recommend to install `shellcheck` locally - or even better the vscode extension if you use vscode. It would highlight these things.

jwendell commented on ossm/ci/pre-submit.sh:
```diff
@@ -0,0 +1,5 @@
+
+DIR=$(cd "$(dirname "$0")" ; pwd -P)
+
+time cargo build --release --features tls-openssl --no-default-features"
```
this file needs `chmod +x`.

zmiklank committed:
do not enter scripts directory, few more nitpicks

Signed-off-by: Zuzana Miklankova <zmiklank@redhat.com>
